Bring forth the template guidelines doc (#741)
Shopify/hydrogen: templates/skeleton/TEMPLATE_GUIDELINES.md

# Hydrogen Template Guidelines

> The following code guidelines are our opinionated best practices for developing templates in Remix and Hydrogen, with the goal of building consistent and well structured code.

Reviewer: Small suggestion, I think we could combine these paragraphs and be slightly more direct. I'm not a great writer, so feel free to take this, adapt it or disregard it entirely.

```suggestion
This initial document provides guidelines and opinionated best practises for developing in Remix and Hydrogen with the goal of building consistent and well structured code. We hope this will evolve into more official recommendations, ESLint presets and other tooling to guide users to the outcomes defined below.
```

Author: ```suggestion
This initial document provides guidelines and opinionated best practices for developing in Remix and Hydrogen with the goal of building consistent and well structured code. This may evolve into more official recommendations, such as ESLint presets and other tooling to guide users to the outcomes defined below.
```

- **Use try/catch** – except in “loader”, “action”, and the Component. Those three “Route Module APIs” are handled automatically by `ErrorBoundary` and CatchBoundary, but the rest – such as “meta”, “links”, “handle”, etc. – will crash the server if an error is thrown.
- **Have a CatchBoundary if necessary.** A CatchBoundary is used when a new Response is thrown, and is generally meant for expected errors caused by the user, such as 401, 404, etc. Note that `CatchBoundary`s will be deprecated in Remix V2, at which time we'll remove this recommendation.

### Don’t:

- **Use try/catch in “loader”, “action”, and the Component.** For templates it’s easier to let the error be thrown and get handled by the `ErrorBoundary` than to handle it manually.

### Examples

PR: https://github.com/Shopify/hydrogen/pull/716

Code:

```tsx
export async function loader() {
  /* failed SFAPI requests will throw to the ErrorBoundary */
  const data = storefront.query()

>   /* If no page data is returned purposely throw to the CatchBoundary */

Author: According to the docs, you’re supposed to throw a Response for the CatchBoundary :shrug:
https://remix.run/docs/en/1.15.0/route/catch-boundary
>A CatchBoundary is a React component that renders whenever an action or loader throws a Response.
>
>Note: We use the word “catch” to represent the codepath taken when a Response type is thrown;

>   /* If no page data is returned purposely throw to the CatchBoundary */

Reviewer: Yeah, agree this is correct and not a point of confusion 👍

  if (!data?.page) {
    throw new Response('Page not found', { status: 404 })
  }

  //...
  return defer()
}

export function meta() {
  try {
    //
  catch(error) {
    //
  }
}

export function ErrorBoundary({error}) {
  return (<div>{error.message}</div>)
}

export function CatchBoundary() {
  const {statusText} = useCatch()
  return (<div>{statusText}</div>)
}

export default function TheUIComponents() {
  return (
    <Await resolve={} errorElement={<div>An error occurred</div>}>
      //
    </Await>
  )
}
```

## Template Dependencies

> Minimalize dependencies and shared code between templates.

Author: We were trying to have the "action item" be the first thing that you read, and then context after that. So what about

```suggestion
Minimalize dependencies and shared code between templates; template files are meant to stand alone in a given Hydrogen storefront. 
```

?

### Do:

- Keep all code associated with a template in a single template file, even if that means some duplication in templates

### Don't:

> - Share code between templates in a separate file. There is no guarantee that any given project will need both those templates, and we also want to avoid issues associated with generating multiple files (such as file placement)

Author: Yeah that came out bad. What I meant was having a helper file that's generated somewhere under `routes/` so that it ends up becoming a Remix route by accident.

> - Share code between templates in a separate file. There is no guarantee that any given project will need both those templates, and we also want to avoid issues associated with generating multiple files (such as file placement)

Author: ```suggestion
- Share code between templates in a separate file. There is no guarantee that any given project will need both those templates, and we also want to avoid issues associated with generating multiple files (such as file organization).
```

<!-- ## Code Organization

Maintain consistency in the styling of the code and the order that things are done within them. This helps developers jump between our templates and know where to look for certain things like `loaders`, etc. -->

<!-- ### Module and Type Imports

Route module imports should follow this proposed order (sorted alphabetically where applicable)

#### Example

```tsx
/* node_modules first */
import invariant from "tiny-invariant";
import { useState } from "react";

/* aliased second */
import { MyComponent } from "~/components";
import { atob, btoa, parser } from "~/lib";

/* local imports third */
import { myFunction } from "./myFunction";

/* type imports last */
import type { Shop } from "@shopify/hydrogen-react/storefront-api-types";
``` -->

## Remix Route APIs

Remix-specific route API functions should be ordered and consistent in style, to help developers quickly scan and find what they're looking for.

### Do:

- Order these APIs following a top-down order of concerns:
  1. Http header tweaks (`shouldRevalidate`, `headers`, `meta`, `links`)
  1. Data manipulation (`loader`, `action`)

>   1. UI (`Route`)

Reviewer 2: not sure about this. should this be the `Component` rather than `Route`?

Reviewer 2: btw this looks better than what I had done! ✅

  1. Error handling (`ErrorBoundary`, `CatchBoundary`)
  1. Storefront API GraphQL query strings
- Use function declarations when possible
- Use the most specific type available for Remix Route APIs.

### Example

```tsx
/* module imports... */
import type {LoaderArgs, ActionArgs} from '@shopify/remix-oxygen';

/* local type defintions */

export async function handle() {}

export async function shouldRevalidate() {}

export async function headers() {}

export async function meta() {}

export async function links() {}

export async function loader({}: LoaderArgs) {}

export async function action({}: ActionArgs) {}

export default function Component() {}

export function ErrorBoundary() {}

export function CatchBoundary() {}

> /* storefront Queries/Mutations */

Reviewer: ```suggestion
/* storefront Queries/Mutations, see more specific recommendations below  */
```

```

## TypeScript

Use the types generated automatically for the Storefront API. This not only demonstrates to devs what these types are and where they can be used, but also helps our repo automatically stay up-to-date with Storefront API changes. It also preserves the human-readable documentation that the Storefront API has on each Storefront object and property.

### Do:

- Use the types from `'@shopify/hydrogen/storefront-api-types'`
- Use TypeScript utilities, such as `Pick`, `Omit`, etc. to format the type into the shape you need
- Use `type` over `interface` when possible
- Use `type` when importing types from packages

### Example

```ts
import type {Product} from '@shopify/hydrogen/storefront-api-types';
type MyCustomProduct = Pick<Product, 'name' | 'description'>;
```

## Remix Loader Return Values

Use the correct return type in `loader()`, `action()`, etc.

### Do:

- Use `json()` by default

> - Use `redirect()` from the `remix` package to redirect

Reviewer: from `remix`? or the remix adaptor?

- Use `defer()` when there is a need to have content streamed in later
- Use `new Response()` for errors (like 404s) and for unique document responses like `.xml` and `.txt`
- Use capitalized and kebab-cased headers in responses, like `Cache-Control`

### Example

```tsx
export async function loader() {
  return json({foo: 'bar'});
}
```

```tsx

> import {redirect} from 'remix';

Reviewer: ```suggestion
import {redirect} from ''@shopify/remix-oxygen';';
```

Correct me if I'm wrong.

export async function loader() {
  return redirect('/');
}
```

```tsx
export async function loader() {
  return json(
    {foo: 'bar'},
    {
      headers: {
        'Cache-Control': 'none',
      },
    },
  );
}
```

## Graphql Query Definitions

> Be consistent in the query's variable name, in addition to the query's name in the query string. This helps the variables to be easily found, and ensure that tooling (like GraphQL Codegen) doesn't get have conflicts in query types and results.

Reviewer: ```suggestion
Be consistent in the query's variable name, in addition to the query's name in the query string. This helps the variables to be easily found, and ensure that tooling (like GraphQL Codegen) doesn't get conflicts in query types and results.
```

Author: ```suggestion
Be consistent in the query's variable name, in addition to the query's name in the query string. This helps the variables to be easily found, and ensure that tooling (like GraphQL Codegen) doesn't have conflicts in query types and results.
```
